panel/src/config/fiber.js

```diff
@@ -0,0 +1,373 @@
+        region.scrollTop  = this.page.scrollRegions[index].top;
+        region.scrollLeft = this.page.scrollRegions[index].left;
+      })
+    }
+  },
+
+  saveScroll() {
+    const regions = Array.from(this.scrollRegions());
+    this.state({
+      ...this.page,
+      scrollRegions: regions.map(region => ({
+        top:  region.scrollTop,
+        left: region.scrollLeft,
+      })),
+    });
+  },
+
+  scrollRegions() {
+    return document.querySelectorAll("[scroll-region]");
+  },
+
+  async setPage(page, { replace = false, preserveScroll = false, preserveState = false } = {}) {
+    // resolve component
+    const component = await this.component(page.$view.component);
+
+    // get all scroll regions
+    page.scrollRegions = page.scrollRegions || [];
+
+    // either replacing the whole state
+    // or pushing onto it
+    if (replace || this.toUrl(page.$url).href === window.location.href) {
+      this.state(page);
+    } else {
+      this.state(page, "push");
+    }
+
+    // swap component
+    let data = clone(page);
+
+    // apply all data
+    data = this.data(data);
+
+    // call and wait for the swap callback
+    await this.swap({ component, page: data, preserveState });
+
+    // reset scrolling if it should not be preserved
+    if (!preserveScroll) {
+      this.resetScroll();
+    }
+  },
+
+  state(page, action = "replace") {
+    this.page = page;
+    window.history[action + "State"](page, "", page.$url);
+  },
+
+  toQuery(search, data) {
+    let params = new URLSearchParams(search);
+
+    // make sure that we always work with a data object
+    if (typeof data !== "object") {
+      data = {};
+    }
+
+    // add all data params unless they are empty/null
+    Object.entries(data).forEach(([key, value]) => {
+      if (value !== null) {
+        params.set(key, value);
+      }
+    });
+
+    return params;
+  },
+
+  toUrl(href, {
+    data = {},
+    hash = true
+  } = {}) {
+    let url
+
+    if (hash === true) {
+      url = new URL(href, window.location);
+    } else {
+      url = new URL(href);
+      url.hash = "";
+    }
+
+    url.search = this.toQuery(url.search, data);
+
+    return url
+  },
+
+  async visit(url, {
+    replace = false,
+    preserveScroll = false,
+    preserveState = false,
+    only = [],
+    headers = {},
+    data = {}
+  } = {}) {
+
+    // save the current scrolling positions
+    // for all scroll regions
+    this.saveScroll();
+
+    document.dispatchEvent(new Event("fiber:start"));
+
+    // make sure only is an array
+    if (Array.isArray(only) === false) {
+      only = [only]
+    }
+
+    // create proper URL
+    url = this.toUrl(url, { data: data || {}, hash: false });
+
+    try {
+      // fetch the response (only GET request supported)
+      const response = await fetch(url, {
+        method: "get",
+        headers: {
+          ...headers,
+          "Accept": "text/html, application/xhtml+xml",
+          "X-Requested-With": "XMLHttpRequest",
```

Comment:
I guess these lines are from Inertia. Do we really need them for our backend logic?

Reply:
I was wondering about it as well. I think we can remove them. Let's keep it as simple as possible.